Python/immutability.c

```diff
@@ -0,0 +1,588 @@
+
+#include "Python.h"
+#include <ctype.h>
+#include <stdbool.h>
+#include <stdio.h>
+#include "pycore_object.h"
+#include "pycore_immutability.h"
+
+
+static PyObject *
+_destroy(PyObject* set, PyObject *objweakref)
+{
+    Py_INCREF(set);
+    if (PySet_Discard(set, objweakref) < 0) {
+        Py_DECREF(set);
+        return NULL;
+    }
+    Py_DECREF(set);
+
+    Py_RETURN_NONE;
+}
+
+static PyMethodDef _destroy_def = {
+    "_destroy", (PyCFunction) _destroy, METH_O
+};
+
+static PyObject *
+type_weakref(struct _Py_immutability_state *state, PyObject *obj)
+{
+    if(state->destroy_cb == NULL){
+        state->destroy_cb = PyCFunction_NewEx(&_destroy_def, state->freezable_types, NULL);
+        if (state->destroy_cb == NULL) {
+            return NULL;
+        }
+    }
+
+    return PyWeakref_NewRef(obj, state->destroy_cb);
+}
+
+static
+int init_state(struct _Py_immutability_state *state)
+{
+    PyObject* frozen_importlib = NULL;
+
+    frozen_importlib = PyImport_ImportModule("_frozen_importlib");
+    if(frozen_importlib == NULL){
+        return -1;
+    }
+
+    state->module_locks = PyObject_GetAttrString(frozen_importlib, "_module_locks");
+    if(state->module_locks == NULL){
+        Py_DECREF(frozen_importlib);
+        return -1;
+    }
+
+    state->blocking_on = PyObject_GetAttrString(frozen_importlib, "_blocking_on");
+    if(state->blocking_on == NULL){
+        Py_DECREF(frozen_importlib);
+        return -1;
+    }
+
+    state->freezable_types = PySet_New(NULL);
+    if(state->freezable_types == NULL){
+        Py_DECREF(frozen_importlib);
+        return -1;
+    }
+
+    if(PyDict_SetItemString(PyModule_GetDict(frozen_importlib), "_freezable_types", state->freezable_types)){
+        Py_DECREF(frozen_importlib);
+        return -1;
+    }
```

Comment:
My understanding is, that this exposes the `freezable_types` set via the `frozen_importlib` module. Is there a reason for this?